src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
-        _pool.withdraw(total);
-        daiJoin.join(address(this), total);
+        // Determine if it needs to unwind or wind
 
-        // normalized debt == erc20 DAI to pool (Vat rate for D3M ilks fixed to 1 RAY)
+        uint256 Line;
+        uint256 debt;
+        uint256 targetAssets;
+        uint256 toUnwind;
 
-        if (mode == Mode.NORMAL) {
-            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, fees * RAY);
-        } else if (mode == Mode.D3M_CULLED) {
-            vat.slip(ilk, address(_pool), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
+        // Determine if it needs to fully unwind due to D3M ilk being caged (but not culled) or plan is not active
+        if (ilks[ilk].tic != 0 || !ilks[ilk].plan.active()) {
+            toUnwind = type(uint256).max; // We make sure to enter the unwind path
         } else {
-            // This can be done with the assumption that the price of 1 collateral unit equals 1 DAI.
-            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
-            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
-            vat.slip(ilk, address(_end), -int256(amount));
-            vat.move(address(this), vow, total * RAY);
-        }
+            Line = vat.Line();
+            debt = vat.debt();
+            targetAssets = ilks[ilk].plan.getTargetAssets(currentAssets);
 
-        emit Unwind(ilk, amount, fees);
-    }
+            // Determine if it needs to unwind due to:
+            unchecked {
+                toUnwind = _max(
+                                _max(
+                                    art > lineWad ? art - lineWad : 0, // ilk debt ceiling exceeded
+                                    debt > Line ? _divup(debt - Line, RAY) : 0 // global debt ceiling exceeded
+                                ),
+                                targetAssets < currentAssets ? currentAssets - targetAssets : 0 // plan targetAssets
+                            );
+            }
+        }
 
-    function _fix(bytes32 ilk, address _pool) internal returns (uint256 diff) {
-        (uint256 ink, uint256 art) = vat.urns(ilk, _pool);
-        if (art < ink) {
-            address _vow = vow;
-            diff = ink - art;
-            require(diff <= MAXINT256, "D3MHub/overflow");
-            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
-            vat.grab(ilk, _pool, _pool, _vow, 0, int256(diff));
+        if (toUnwind > 0) {
+            toUnwind = _min(
+                toUnwind,
+                maxWithdraw
+            );
+            require(toUnwind <= MAXINT256, "D3MHub/overflow");
```

Comment:
This check should be moved inside of `_unwind`.

Reply:
Yeah I guess this is subjective, we did this way to keep very symmetric `_unwind` from `_wind`, and the math to check in `_wind` is a bit more complex. That's the main reason to have it extracted out.
We could even argue if maybe these two internal should just be incorporated to `_normal` as they are called just once and shouldn't provoke stack too deep either. Actually thinking seriously about this now.

Reply:
I've removed the regular `_wind` and `_unwind` and put the code in the main internal function, which is called `_exec` now. And the full `_unwind` is called `_wipe` now.

Reply:
@kmbarry1 let me know if we can resolve this one.